pkg/integrations/v4/dm/emitter.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2020 New Relic Corporation. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package dm
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/newrelic/infrastructure-agent/internal/agent"
+	"github.com/newrelic/infrastructure-agent/internal/agent/cmdchannel/handler"
+	"github.com/newrelic/infrastructure-agent/internal/feature_flags"
+	"github.com/newrelic/infrastructure-agent/internal/integrations/v4/integration"
+	"github.com/newrelic/infrastructure-agent/pkg/backend/http"
+	"github.com/newrelic/infrastructure-agent/pkg/backend/identityapi"
+	"github.com/newrelic/infrastructure-agent/pkg/databind/pkg/data"
+	"github.com/newrelic/infrastructure-agent/pkg/entity"
+	"github.com/newrelic/infrastructure-agent/pkg/integrations/legacy"
+	"github.com/newrelic/infrastructure-agent/pkg/integrations/v4/protocol"
+	"github.com/newrelic/infrastructure-agent/pkg/log"
+)
+
+var (
+	// Errors
+	ProtocolV4NotEnabledErr = errors.New("integration protocol version 4 is not enabled")
+	NoContentToParseErr     = errors.New("no content to parse")
+
+	// internal
+	elog = log.WithComponent("DimensionalMetricsEmitter")
+)
+
+const (
+	nrEntityId = "nr.entity.id"
+)
+
+type Agent interface {
+	GetContext() agent.AgentContext
+}
+
+type emitter struct {
+	ffRetriever    feature_flags.Retriever
+	metricsSender  MetricsSender
+	agentContext   agent.AgentContext
+	registerClient identityapi.RegisterClient
+}
+
+type Emitter interface {
+	Send(
+		metadata integration.Definition,
+		extraLabels data.Map,
+		entityRewrite []data.EntityRewrite,
+		integrationJSON []byte) error
+}
+
+func NewEmitter(
+	agentContext agent.AgentContext,
+	dmSender MetricsSender,
+	ffRetriever feature_flags.Retriever,
+	registerClient identityapi.RegisterClient) Emitter {
+
+	return &emitter{
+		agentContext:   agentContext,
+		metricsSender:  dmSender,
+		ffRetriever:    ffRetriever,
+		registerClient: registerClient,
+	}
+}
+
+func (e *emitter) Send(
+	metadata integration.Definition,
+	extraLabels data.Map,
+	entityRewrite []data.EntityRewrite,
+	integrationJSON []byte) error {
+
+	pluginDataV4, err := ParsePayloadV4(integrationJSON, e.ffRetriever)
+	if err != nil {
+		elog.WithError(err).WithField("output", string(integrationJSON)).Warn("can't parse v4 integration output")
```

Comment:
Not sure about the parser being responsable of FF handling or even the emitter. 

Is it possible extract it from here to an up level?

I'm thinking to have the emitter agnostic w/o FF logic and leave that responsibility on who is call it.

Not sure if it make sense.

Reply:
Agree, this should probably not live here. This was from the old `emitter_v4.go` code that I moved. I'd like to move/refactor it out but trying to make minimum "big" changes.